tao_compiler/mlir/disc/transforms/disc_algebraic_simplifier.cc

```diff
@@ -530,6 +533,93 @@ struct IndexCastSimplifierPattern
     return failure();
   }
 };
+// Simplify get_dimension_size pattern. An examples as following:
+// Case 1):
+// %2 = "mhlo.get_dimension_size"(%1)
+// %3 = "tensor.extract" %2[] -> i32
+// %from_elements = tensor.from_elements %3, ...
+// Convert to:
+// %2 = "tensor.dim"(%1, %cst0) -> i32
+// %from_elements = tensor.from_elements %2, ...
+//
+// Case 2):
+// %2 = "mhlo.get_dimension_size"(%1)
+// %3 = mhlo.mul %2, %4
+// %4 = "tensor.extract" %3[] -> i32
+// %from_elements = tensor.from_elements %4, ...
+// Convert to:
+// %2 = "tensor.dim"(%1, %cst0) -> i32
+// %3 = arith.mul %2, %4
+// %from_elements = tensor.from_elements %3, ...
+
+struct SimplifierGetDimensionSizePattern
+    : public OpRewritePattern<mhlo::GetDimensionSizeOp> {
+  using OpRewritePattern<mhlo::GetDimensionSizeOp>::OpRewritePattern;
+  LogicalResult matchAndRewrite(mhlo::GetDimensionSizeOp getDimOp,
+                                PatternRewriter& rewriter) const override {
+    auto loc = getDimOp->getLoc();
+    Value tensor = getDimOp->getOperand(0);
+    auto dim = getDimOp.getDimension();
+    auto elemTy = getDimOp.getResult()
+                      .getType()
+                      .cast<RankedTensorType>()
+                      .getElementType();
+
+    SmallVector<Operation*, 4> ops;
+    std::stack<Operation*> stack;
+    for (auto user : getDimOp->getUsers()) {
+      if (isa<tensor::ExtractOp, func::ReturnOp, mhlo::ReshapeOp,
+              mhlo::DynamicBroadcastInDimOp>(user))
+        continue;
+      stack.push(user);
+    }
+    while (!stack.empty()) {
+      auto user = stack.top();
+      stack.pop();
+      ops.push_back(user);
+      for (auto op : user->getUsers()) {
+        if (isa<tensor::ExtractOp, func::ReturnOp, mhlo::ReshapeOp,
+                mhlo::DynamicBroadcastInDimOp>(op))
+          continue;
+        stack.push(op);
+      }
+    }
+    for (auto op : ops) {
+      auto loc = op->getLoc();
+      rewriter.setInsertionPoint(op);
+      auto v1 = rewriter.create<tensor::ExtractOp>(loc, op->getOperand(0));
+      auto v2 = rewriter.create<tensor::ExtractOp>(loc, op->getOperand(1));
+      Value newOpValue;
+      if (isa<mhlo::MulOp>(op)) {
+        newOpValue = rewriter.create<arith::MulIOp>(loc, v1, v2).getResult();
+      } else if (isa<mhlo::AddOp>(op)) {
+        newOpValue = rewriter.create<arith::AddIOp>(loc, v1, v2).getResult();
+      } else if (isa<mhlo::SubtractOp>(op)) {
+        newOpValue = rewriter.create<arith::SubIOp>(loc, v1, v2).getResult();
+      } else if (isa<mhlo::DivOp>(op)) {
+        newOpValue = rewriter.create<arith::DivSIOp>(loc, v1, v2).getResult();
+      } else {
+        return failure();
```

Comment:
nit: just skip this op or fail this pass?

Reply:
Failure is better, let's rewrite MHLO operations using arithmetic operations to compute scalar values.